### ISSUE TITLE
Every apt command fails with "Could not get lock ... Temporarily unavailable"

### ISSUE THREAD

**Issue by mikker** [progrium/dokku#1205]
I'm not sure when it happened but something seems to have gone awry with my dokku setup and now every apt command fails eg:

```
$ apt-get update && apt-get upgrade -y
Ign http://ppa.launchpad.net trusty InRelease
Hit http://ppa.launchpad.net trusty Release.gpg
Hit http://ppa.launchpad.net trusty Release
Hit http://ppa.launchpad.net trusty/main amd64 Packages
Hit http://ppa.launchpad.net trusty/main i386 Packages
Hit http://ppa.launchpad.net trusty/main Translation-en
Ign http://security.ubuntu.com trusty-security InRelease
Ign http://mirrors.digitalocean.com trusty InRelease
Get:1 http://security.ubuntu.com trusty-security Release.gpg [933 B]
Ign http://mirrors.digitalocean.com trusty-updates InRelease
Get:2 http://security.ubuntu.com trusty-security Release [63.5 kB]
Hit http://mirrors.digitalocean.com trusty Release.gpg
Hit http://mirrors.digitalocean.com trusty-updates Release.gpg
Hit http://mirrors.digitalocean.com trusty Release
Hit http://mirrors.digitalocean.com trusty-updates Release
Get:3 http://security.ubuntu.com trusty-security/main Sources [81.4 kB]
Hit http://mirrors.digitalocean.com trusty/main Sources
Hit http://mirrors.digitalocean.com trusty/universe Sources
Hit http://mirrors.digitalocean.com trusty/main amd64 Packages
Get:4 http://security.ubuntu.com trusty-security/universe Sources [25.2 kB]
Hit http://mirrors.digitalocean.com trusty/universe amd64 Packages
Get:5 http://security.ubuntu.com trusty-security/main amd64 Packages [272 kB]
Hit http://mirrors.digitalocean.com trusty/main i386 Packages
Get:6 http://get.docker.io docker InRelease
Hit http://mirrors.digitalocean.com trusty/universe i386 Packages
Ign http://get.docker.io docker InRelease
Hit http://mirrors.digitalocean.com trusty/main Translation-en
Get:7 http://security.ubuntu.com trusty-security/universe amd64 Packages [104 kB]
Hit http://mirrors.digitalocean.com trusty/universe Translation-en
Get:8 http://security.ubuntu.com trusty-security/main i386 Packages [260 kB]
Hit http://mirrors.digitalocean.com trusty-updates/main Sources
Hit http://mirrors.digitalocean.com trusty-updates/universe Sources
Get:9 http://security.ubuntu.com trusty-security/universe i386 Packages [104 kB]
Hit https://packagecloud.io trusty InRelease
Hit http://mirrors.digitalocean.com trusty-updates/main amd64 Packages
Hit http://mirrors.digitalocean.com trusty-updates/universe amd64 Packages
Hit http://security.ubuntu.com trusty-security/main Translation-en
Hit http://mirrors.digitalocean.com trusty-updates/main i386 Packages
Hit http://security.ubuntu.com trusty-security/universe Translation-en
Hit http://get.docker.io docker Release.gpg
Hit http://mirrors.digitalocean.com trusty-updates/universe i386 Packages
Hit http://mirrors.digitalocean.com trusty-updates/main Translation-en
Hit https://packagecloud.io trusty/main amd64 Packages
Hit http://mirrors.digitalocean.com trusty-updates/universe Translation-en
Hit https://packagecloud.io trusty/main i386 Packages
Get:10 https://packagecloud.io trusty/main Translation-en_US
Hit http://get.docker.io docker Release
Ign http://mirrors.digitalocean.com trusty/main Translation-en_US
Hit http://get.docker.io docker/main amd64 Packages
Ign http://mirrors.digitalocean.com trusty/main Translation-en_US.UTF-8
Ign http://mirrors.digitalocean.com trusty/universe Translation-en_US
Ign http://mirrors.digitalocean.com trusty/universe Translation-en_US.UTF-8
Hit http://get.docker.io docker/main i386 Packages
Ign https://packagecloud.io trusty/main Translation-en_US
Ign https://packagecloud.io trusty/main Translation-en
Ign https://packagecloud.io trusty/main Translation-en_US.UTF-8
Get:11 http://get.docker.io docker/main Translation-en_US
Ign http://get.docker.io docker/main Translation-en_US
Ign http://get.docker.io docker/main Translation-en
Ign http://get.docker.io docker/main Translation-en_US.UTF-8
Fetched 911 kB in 24s (37.9 kB/s)
^Cading package lists... 99%
root@brnbw dokku $ apt-get update && apt-get upgrade -y -V
Ign http://ppa.launchpad.net trusty InRelease
Hit http://ppa.launchpad.net trusty Release.gpg
Hit http://ppa.launchpad.net trusty Release
Hit http://ppa.launchpad.net trusty/main amd64 Packages
Hit http://ppa.launchpad.net trusty/main i386 Packages
Hit http://ppa.launchpad.net trusty/main Translation-en
Ign http://security.ubuntu.com trusty-security InRelease
Ign http://mirrors.digitalocean.com trusty InRelease
Hit http://security.ubuntu.com trusty-security Release.gpg
Ign http://mirrors.digitalocean.com trusty-updates InRelease
Hit http://security.ubuntu.com trusty-security Release
Hit http://mirrors.digitalocean.com trusty Release.gpg
Hit http://security.ubuntu.com trusty-security/main Sources
Hit http://security.ubuntu.com trusty-security/universe Sources
Hit http://mirrors.digitalocean.com trusty-updates Release.gpg
Hit http://security.ubuntu.com trusty-security/main amd64 Packages
Hit http://mirrors.digitalocean.com trusty Release
Hit http://security.ubuntu.com trusty-security/universe amd64 Packages
Hit http://mirrors.digitalocean.com trusty-updates Release
Hit http://security.ubuntu.com trusty-security/main i386 Packages
Hit http://mirrors.digitalocean.com trusty/main Sources
Hit http://security.ubuntu.com trusty-security/universe i386 Packages
Hit http://security.ubuntu.com trusty-security/main Translation-en
Hit http://mirrors.digitalocean.com trusty/universe Sources
Hit http://security.ubuntu.com trusty-security/universe Translation-en
Hit http://mirrors.digitalocean.com trusty/main amd64 Packages
Hit http://mirrors.digitalocean.com trusty/universe amd64 Packages
Get:1 http://get.docker.io docker InRelease
Hit http://mirrors.digitalocean.com trusty/main i386 Packages
Hit http://mirrors.digitalocean.com trusty/universe i386 Packages
Ign http://get.docker.io docker InRelease
Hit http://mirrors.digitalocean.com trusty/main Translation-en
Hit http://mirrors.digitalocean.com trusty/universe Translation-en
Hit http://mirrors.digitalocean.com trusty-updates/main Sources
Hit http://mirrors.digitalocean.com trusty-updates/universe Sources
Hit http://get.docker.io docker Release.gpg
Hit http://mirrors.digitalocean.com trusty-updates/main amd64 Packages
Hit http://mirrors.digitalocean.com trusty-updates/universe amd64 Packages
Hit http://mirrors.digitalocean.com trusty-updates/main i386 Packages
Hit https://packagecloud.io trusty InRelease
Hit http://mirrors.digitalocean.com trusty-updates/universe i386 Packages
Hit http://mirrors.digitalocean.com trusty-updates/main Translation-en
Hit http://mirrors.digitalocean.com trusty-updates/universe Translation-en
Hit https://packagecloud.io trusty/main amd64 Packages
Hit http://get.docker.io docker Release
Hit https://packagecloud.io trusty/main i386 Packages
Get:2 https://packagecloud.io trusty/main Translation-en_US
Hit http://get.docker.io docker/main amd64 Packages
Ign http://mirrors.digitalocean.com trusty/main Translation-en_US
Ign http://mirrors.digitalocean.com trusty/main Translation-en_US.UTF-8
Hit http://get.docker.io docker/main i386 Packages
Ign http://mirrors.digitalocean.com trusty/universe Translation-en_US
Ign http://mirrors.digitalocean.com trusty/universe Translation-en_US.UTF-8
Get:3 http://get.docker.io docker/main Translation-en_US
Ign https://packagecloud.io trusty/main Translation-en_US
Ign https://packagecloud.io trusty/main Translation-en
Ign https://packagecloud.io trusty/main Translation-en_US.UTF-8
Ign http://get.docker.io docker/main Translation-en_US
Ign http://get.docker.io docker/main Translation-en
Ign http://get.docker.io docker/main Translation-en_US.UTF-8
Reading package lists... Done
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
The following packages have been kept back:
   linux-generic (3.13.0.52.59 => 3.13.0.53.60)
   linux-headers-generic (3.13.0.52.59 => 3.13.0.53.60)
   linux-image-generic (3.13.0.52.59 => 3.13.0.53.60)
   lxc-docker (1.6.1 => 1.6.2)
The following packages will be upgraded:
   dokku (0.3.18 => 0.3.18)
1 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.
1 not fully installed or removed.
Need to get 21.6 kB of archives.
After this operation, 0 B of additional disk space will be used.
Fetched 21.6 kB in 2s (9,434 B/s)
Preconfiguring packages ...
(Reading database ... 97660 files and directories currently installed.)
Preparing to unpack .../dokku_0.3.18_amd64.deb ...
Unpacking dokku (0.3.18) over (0.3.18) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Setting up dokku (0.3.18) ...
Purging old database entries in /usr/share/man...
Processing manual pages under /usr/share/man...
Purging old database entries in /usr/share/man/fr.UTF-8...
Processing manual pages under /usr/share/man/fr.UTF-8...
Purging old database entries in /usr/share/man/tr...
Processing manual pages under /usr/share/man/tr...
Purging old database entries in /usr/share/man/fi...
Processing manual pages under /usr/share/man/fi...
Purging old database entries in /usr/share/man/ko...
Processing manual pages under /usr/share/man/ko...
Purging old database entries in /usr/share/man/da...
Processing manual pages under /usr/share/man/da...
Purging old database entries in /usr/share/man/gl...
Processing manual pages under /usr/share/man/gl...
Purging old database entries in /usr/share/man/pt_BR...
Processing manual pages under /usr/share/man/pt_BR...
Purging old database entries in /usr/share/man/ro...
Processing manual pages under /usr/share/man/ro...
Purging old database entries in /usr/share/man/hu...
Processing manual pages under /usr/share/man/hu...
Purging old database entries in /usr/share/man/vi...
Processing manual pages under /usr/share/man/vi...
Purging old database entries in /usr/share/man/id...
Processing manual pages under /usr/share/man/id...
Purging old database entries in /usr/share/man/pl...
Processing manual pages under /usr/share/man/pl...
Purging old database entries in /usr/share/man/uk...
Processing manual pages under /usr/share/man/uk...
Purging old database entries in /usr/share/man/sv...
Processing manual pages under /usr/share/man/sv...
Purging old database entries in /usr/share/man/fr...
Processing manual pages under /usr/share/man/fr...
Purging old database entries in /usr/share/man/zh...
Processing manual pages under /usr/share/man/zh...
Purging old database entries in /usr/share/man/sl...
Processing manual pages under /usr/share/man/sl...
Purging old database entries in /usr/share/man/es...
Processing manual pages under /usr/share/man/es...
Purging old database entries in /usr/share/man/it...
Processing manual pages under /usr/share/man/it...
Purging old database entries in /usr/share/man/eo...
Processing manual pages under /usr/share/man/eo...
Purging old database entries in /usr/share/man/fr.ISO8859-1...
Processing manual pages under /usr/share/man/fr.ISO8859-1...
Purging old database entries in /usr/share/man/pt...
Processing manual pages under /usr/share/man/pt...
Purging old database entries in /usr/share/man/sk...
Processing manual pages under /usr/share/man/sk...
Purging old database entries in /usr/share/man/ja...
Processing manual pages under /usr/share/man/ja...
Purging old database entries in /usr/share/man/de...
Processing manual pages under /usr/share/man/de...
Purging old database entries in /usr/share/man/hr...
Processing manual pages under /usr/share/man/hr...
Purging old database entries in /usr/share/man/zh_CN...
Processing manual pages under /usr/share/man/zh_CN...
Purging old database entries in /usr/share/man/zh_TW...
Processing manual pages under /usr/share/man/zh_TW...
Purging old database entries in /usr/share/man/cs...
Processing manual pages under /usr/share/man/cs...
Purging old database entries in /usr/share/man/sr...
Processing manual pages under /usr/share/man/sr...
Purging old database entries in /usr/share/man/el...
Processing manual pages under /usr/share/man/el...
Purging old database entries in /usr/share/man/ru...
Processing manual pages under /usr/share/man/ru...
Purging old database entries in /usr/share/man/nl...
Processing manual pages under /usr/share/man/nl...
Purging old database entries in /usr/local/share/man...
Processing manual pages under /usr/local/share/man...
0 man subdirectories contained newer manual pages.
0 manual pages were added.
0 stray cats were added.
0 old database entries were purged.
User 'dokku' already exists
docker:x:999:dokku
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
dpkg: error processing package dokku (--configure):
 subprocess installed post-installation script returned error exit status 100
Errors were encountered while processing:
 dokku
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

```
$ docker version
Client version: 1.6.1
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 97cd073
OS/Arch (client): linux/amd64
Server version: 1.6.1
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 97cd073
OS/Arch (server): linux/amd64
$ dokku version
0.3.18
```

Can I circumvent the user creation in any way or is this in fact a bug?

**Comment by mikker** [comment]
No idea what fixed it but it's gone. And User exists definitely wasn't what was really wrong. Forget I ever asked.

**Comment by matto1990** [comment]
I'm getting the same issue as this. It seems to be an issue with the dpkg lock, but I can't figure out why this would be the case. any help?

**Comment by mikker** [comment]
I got around it by uninstalling then installing on top of my old config (and then wrestle a bit). But that's clearly not a good solution. I'm unsure but I think I started with a DigitalOcean VPS that had Dokku preinstalled AND then installed dokku on top. But I can't really remember to be honest. @matto1990 did you start with the dokku droplet?

**Comment by matto1990** [comment]
No I started with a basic DO Ubuntu 14.04 x64 droplet and installed 0.3.18 using apt-get. I got this error while trying to upgrade for the first time to 0.3.21. It seems to be an issue in the configuration part of the upgrade. Now I've upgraded it once it will show this error anytime I use apt-get, even with other pages, as it tries and fails to configure the dokku package again each time.

So to be clear you:

- Uninstalled the dokku package: `sudo apt-get remove dokku`
- Then installed it again: `sudo apt-get install dokku`

?

**Comment by mikker** [comment]
Yes, that's what I did. And `buildstep`. `sudo apt-get remove dokku buildstep`. I think I also backed up my plugins beforehand.

**Comment by matto1990** [comment]
Does that remove any of the apps or will they still be there once the installation is done?

**Comment by mikker** [comment]
Mine were still there. As far as I understand it (and I barely understand anything) they're owned by docker, so as long as you don't uninstall they'll be left where they are. But you know; no promises.

**Comment by matto1990** [comment]
Ok. I'll give it a go.

_fingers crossed_

**Comment by mikker** [comment]
It got me to 0.3.21 but I still can't update from there.

**Comment by matto1990** [comment]
That didn't work for me. I'm going back to a backup from before I started trying to upgrade to try again.

Thanks for the help!

**Comment by matto1990** [comment]
I'm still getting the same issue with the backup. Can anyone provide any help for this?

**Comment by michaelshobbs** [comment]
We had a release issue at some point recently. Does uninstalling/installing still not work?

**Comment by mikker** [comment]
@michaelschobbs Tried today. It still doesn't work.

**Comment by mikker** [comment]
Sorry, @michaelshobbs.

**Comment by michaelshobbs** [comment]
@josegonzalez halp?

**Comment by michaelshobbs** [comment]
If this is the real error then there's nothing for dokku to do. This is apt/environment issue.

```
E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?
```

Check for cronjobs or other users on the system running `apt` command.

**Comment by mikker** [comment]
I think it is – but it only occurs during the upgrade of dokku.

**Comment by 3onyc** [comment]
Do you have any additional plugins installed? Some plugins install additional packages during `dokku plugins-install`, which runs when dokku is upgraded. (See [debian/postinst#L19](https://github.com/progrium/dokku/blob/master/debian/postinst#L19)).

Try temporarily removing any plugins you added and re-running `apt-get upgrade`

**Comment by michaelshobbs** [comment]
Good call @3onyc. @mikker try capturing `ps auxwww`a few times while running the upgrade. with that you should be able to see what process is spawning another apt process.

**Comment by mikker** [comment]
Holy :hankey:. I did that and it turns out, the plugin causing trouble was one I've made myself! Turns out. Sorry to misdirect you so much and thanks a lot for finally leading me there.

**Comment by 3onyc** [comment]
It's quite common for plug-ins to install packages through apt-get, any idea what we can do to either circumvent this issue or offer a better alternative to plugin developers?

// ping @michaelshobbs

**Comment by mikker** [comment]
I was trying to install curl in my `install` script. Ideally it would be possible to supply a list of dependencies. But then again that's a whole other set of problems.

**Comment by josegonzalez** [comment]
Interesting, hadn't thought of that when I added the `install` phase in to the debian package. Probably worth removing and telling people to run `install` manually...

**Comment by 3onyc** [comment]
Preferably people would create their plugins as Debian packages, that solves installing dependencies too.

@josegonzalez Sounds like the best solution yeah, perhaps with a message telling them to run `dokku plugins-install` afterwards.

**Comment by matto1990** [comment]
I know this wasn't an issue with Dokku itself, but shouldn't this be left open until the workaround suggested by @josegonzalez and @3onyc is implemented?

**Comment by michaelshobbs** [comment]
re-opening now because (I forgot and) I think #1387 is related

**Comment by jimeh** [comment]
I came across this from #1387. I quite like how you can easily install plugins by simply dropping them into a folder, makes it especially easy when you're in need of some quick custom plugins. So I'm not a huge fan of the idea to require plugins to be apt packages.

So here's my idea:

1. Have the deb post-install script only execute the `install` hook from built-in plugins.
2. Show some form of post install warning/info to manually run `dokku plugins-install` to ensure all user installed plugins are operational.

This means all existing plugins can keep working without modification, and I believe asking to run one separate command as part of the upgrade process is not a massive deal.

Simplest way to implement this would probably be to move built-in plugins from `/var/lib/dokku/plugins`, leaving that folder to only be used for custom plugins. Then the post-install script could simply execute plugins from the built-in folder only, while the normal `dokku plugins-install` command would run against both plugin folders.

And I have another idea which might be a bit worse/stupid, but I'll throw it out there regardless. What about deprecating the `install` hook in plugins in favor of some convention to trigger setup/installation steps only if needed when specific hooks are called. For example plugins could define a `setup_plugin_name` bash function that's called at the beginning of each hook.

Personally I'm using chef and the [simple-dokku](https://github.com/rlister/chef-dokku-simple) cookbook to manage Dokku. My solution for now will be to update the cookbook to remove all plugins it manages before upgrading dokku, and then reinstalling them all again. It's an alright solution for automated setups, but obviously quite annoying for anyone manually managing dokku.

**Comment by michaelshobbs** [comment]
I don't think we can feasibly require folks to write debs for plugins. Part of the benefit (IMO) of dokku and the plugin ecosystem is that plugin authoring is *dead* simple. Adding a deb package requirement severely hampers that.

Currently we use [pluginhook](https://github.com/progrium/pluginhook) for plugin execution and it only supports a single `PLUGIN_PATH`. So supporting two paths is not currently possible nor on the roadmap. We will be moving to [plugn](https://github.com/progrium/plugn) at some point (#1228) but that still does not support multiple plugin paths.

Regarding the deprecation of the `install` plugin hook, I think you can get the same end result by writing more logic into your `install` script.

All that said, the root of this issue is that `apt-get install dokku` ends up calling `dokku plugins-install` and certain plugins attempt to run `apt-get install` as part of their `install` hooks. Hence the lock error. Unfortunately, I think the only way around this issue is to decouple `plugins-install` from the dokku package installation and pop a message that reminds the user to run it manually.

**Comment by jimeh** [comment]
Exactly, I think executing the `install` hook of custom plugins as part of the deb post-install script needs to be removed. As far as I can tell there's four core plugins with install scripts, and as they're part of dokku it makes sense for them to run as part of the package installation/update. In theory core plugins can be kept free of any `apt-get install` calls, as they can rely on Dokku's own installation method providing any packages they need.

Meaning dokku itself will work after an installation/upgrade, but if you've got custom plugins, you should also run `dokku plugins-install` to ensure your custom plugins are setup correctly.

For a technical means of separating and only running core plugins, I'd imagine pluginhook would need updating if looking for `.core` files were to be the chosen solution.

So instead to keep things simple, how about having a separate folder where core plugins are installed, with each core plugin symlinked into the regular `/var/lib/dokku/plugins` folder? The deb post-install script could set `PLUGINS_PATH` to the dir containing only core plugins, while normal CLI calls to `dokku plugins-install` would set it to `/var/lib/dokku/plugins` allowing core and custom plugins to work at the same time.

**Comment by michaelshobbs** [comment]
The least complex thing to do would be to remove the call to `install` from the deb package and document the need to run `dokku plugins-install` manually.

**Comment by jimeh** [comment]
lol... that is true. But it would sacrifice the vanilla dokku installation experience, as at least without modifying the core plugins that use the `install` hook, dokku wouldn't be properly installed until the user also runs `dokku plugins-install`. Rather than only requiring this step when the user has manually added plugins and hence is required to do so due to their own customizations. But it's probably not a big deal if it's just added to the installation instructions page. Either way, anything is better than the current situation of breaking apt-get :)

**Comment by michaelshobbs** [comment]
Bah! The real issue is calling `install` during upgrade anyway when folks have already dropped 3rd party plugins in place.

@josegonzalez you're the deb master what do you think?

**Comment by josegonzalez** [comment]
Lets just drop that. I can do so and make a release for that today.

**Comment by josegonzalez** [comment]
Closed by 1bf30bdcb1d364e8356bd1a2bc8955d0c34e4409.

**Comment by josegonzalez** [comment]
Reopened as the commit was reverted.

**Comment by josegonzalez** [comment]
One thing I noticed is that for all the new service plugins, we will need to handle not running `dokku plugins-install`. We *could* provide a helper function to detect whether dokku is updating, and then let plugins use that function to optionally run any package management stuff. Here is what I *think* will work, but haven't tested:

```
is_dokku_updating() {
  DOKKU_UPDATING=1
  PIDS=( "$(pstree -Asp $$ | grep -o '([0-9]\+)' | grep -o '[0-9]\+')" )
  for pid in $PIDS; do
    if [[ ! -f "/proc/$pid/cmdline" ]]; then
      continue
    fi
    if grep -q -e "^apt-get" "/proc/$pid/cmdline"; then
      DOKKU_UPDATING=0
      break
    fi
  done
  return "$DOKKU_UPDATING"
}
```

**Comment by Flink** [comment]
Well, I just read the entire discussion and I have an idea which could be relatively simple if it works as I think. We could add a `--core` flag to `dokku plugins-install` which then tells dokku to only install core plugins.
How do we do that? Well by implementing another command which disable a provided plugin (say `dokku plugins:disable <name>`). To disable a plugin we could juste create a directory (named `disabled-plugins`) into `/var/lib/dokku/` and move the provided plugin there. So when calling `dokku plugins-install --core` we just have to disable all plugins not marked with a `.core` file, do the install and then re-enable them. It’s not an atomic operation and there are maybe flaws but it could work, what do you think?

@josegonzalez your function seems quite useful but I don’t know if it will be needed with what I just described?

**Comment by josegonzalez** [comment]
Pluginhook doesn't conditionally trigger hooks based on the presence of a file, so we can't do that. We'd have to reimplement pluginhook for this specific case...

**Comment by michaelshobbs** [comment]
side note: plugn supports disabling/enabling plugins

**Comment by Flink** [comment]
@josegonzalez hmmm I think I wasn’t very clear, what I propose is to physically move the plugins out of the plugin path. So pluginhook couldn’t execute anything, plugins won’t be there anymore :)

**Comment by Flink** [comment]
@michaelshobbs good to know!

**Comment by jimeh** [comment]
Being able to enable/disable plugins is a feature that would be nice to have I believe. But I still think it'd be simpler/cleaner to have a separate directory where core plugins are installed with each of them symlinked to the default `/var/lib/dokku/plugins` path.

This way `dokku plugins-install --core` could easily be enabled by just setting `PLUGINS_PATH` to the directory that holds the core plugins. While without the `--core` flag it'd be set to `/var/lib/dokku/plugins` containing symlinks to all core plugins, and any custom plugins the user has added.

It would also make plugin management a bit easier, as anything symlinked to the core dir is obviously a core plugin, anything else is custom stuff.

**Comment by Flink** [comment]
There this line `export PLUGIN_PATH=${PLUGIN_PATH:="/var/lib/dokku/plugins"}` in the `dokku` script, so I think what you propose @jimeh is totally doable. I may give it a try via a PR.

**Comment by josegonzalez** [comment]
That *could* work, though it sounds wonky. I guess make the PR and we'll see how it looks.

Note: this will require that we do something for upgraded dokku installs to remove the old core plugins or weird stuff might happen.

**Comment by Flink** [comment]
Yes it seems more a hack than a proper solution but still it might be not that bad :) I’ll try and let you know!

**Comment by jimeh** [comment]
@Flink Cool, I'd been meaning to have a closer look at the codebase myself and try and do it, but I haven't had time yet. May I suggest core plugins are installed to a directory named `core-plugins` or `builtin-plugins` within `/var/lib/dokku`, to follow your suggested convention of `disabled-plugins` :)

@josegonzalez I was about to mention that upgrades will need move and symlink plugins into the correct place.

The hack-level of the symlink solution can be argued, but personally I think it's a decently clean solution, as it also physically separates core plugins essential to dokku, from 3rd party custom plugins. It always felt weird that I was adding my random plugins next to core ones. Internal stuff should be internal.

**Comment by Flink** [comment]
@jimeh since plugins path will change I’m going with `/var/lib/dokku/{core,all,disabled}` :)

**Comment by jimeh** [comment]
@Flink Or alternatively if the enable/disabled command comes along too, how about `core`, `custom`, and `enabled`. It would mean you install custom plugins to `custom` and then have to run `dokku plugins:enable <name>` before it'd work. In that way it'd be more like nginx's `sites-{available,enabled}` thing. Not sure it's a great idea, but just thought I'd throw it out there... lol :)

**Comment by josegonzalez** [comment]
Why are we doing subdirectories?

**Comment by Flink** [comment]
@jimeh yep, I’m doing this, I also thought about nginx :D

@josegonzalez I’m trying something with symlinks :)

**Comment by josegonzalez** [comment]
Closing this as there are multiple PRs and solutions to the problem.

Something I've done in the official datastore plugins is to check whether binaries/deps are installed, otherwise output a message to users telling them to run the specific `plugins` command. Works well enough.
